device/cuda/src/clusterization/component_connection.cu

```diff
@@ -0,0 +1,86 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Project include(s).
+#include "component_connection.hpp"
+#include "traccc/cuda/utils/definitions.hpp"
+
+namespace traccc::cuda {
+
+namespace kernel {
+__global__ void component_connection(
+    cell_container_types::const_view cells_view,
+    vecmem::data::vector_view<std::size_t> cluster_prefix_sum_view,
+    cluster_container_types::view clusters_view,
+    vecmem::data::jagged_vector_view<unsigned int> sparse_ccl_indices_view,
+    vecmem::data::vector_view<const device::prefix_sum_element_t>
+        cells_prefix_sum_view) {
+    auto idx = threadIdx.x + blockDim.x * blockIdx.x;
+
+    // Get device vector of the cells prefix sum
+    vecmem::device_vector<const device::prefix_sum_element_t> cells_prefix_sum(
+        cells_prefix_sum_view);
+    // Ignore if idx is out of range
+    if (idx >= cells_prefix_sum.size())
+        return;
+
+    // Get the indices for the module idx and the cell idx
+    auto module_idx = cells_prefix_sum[idx].first;
+    auto cell_idx = cells_prefix_sum[idx].second;
+
+    // Initialize the device containers for cells and clusters
+    cell_container_types::const_device cells_device(cells_view);
+    cluster_container_types::device clusters_device(clusters_view);
+
+    // Get the cells for the current module idx
+    const auto& cells = cells_device.at(module_idx).items;
+
+    // Vectors used for cluster indices found by sparse CCL
+    vecmem::jagged_device_vector<unsigned int> device_sparse_ccl_indices(
+        sparse_ccl_indices_view);
+    const auto& cluster_indices = device_sparse_ccl_indices.at(module_idx);
+
+    // Number of clusters found for this module
+    const auto num_clusters = cluster_indices.back();
```

Comment:
I might be confusing things here, but we previously performed a prefix sum with non-deterministic order, so I cannot help but think that ensuring that the total sum is at the end of the vector is a bit odd?

Reply:
Yes, I agree that this was a little confusing. `cluster_indices` is not the prefix sum vector here. This comes from the logic that we put the number of clusters per module at the last, extra place in each inner vector with sparse_ccl indices.

Similarly, as I mentioned above, @Chamodya-ka, you can have a look at #205, where this confusing logic is gone.

Reply:
Thanks alot! Yea I did come across this while refactoring the code with #205.